components/swim/src/server/outbound.rs

```diff
@@ -0,0 +1,410 @@
+    Ping,
+    PingReq,
+}
+
+impl fmt::Display for AckFrom {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &AckFrom::Ping => write!(f, "Ping"),
+            &AckFrom::PingReq => write!(f, "PingReq"),
+        }
+    }
+}
+
+/// The timing of the outbound thread.
+#[derive(Debug, Clone)]
+pub struct Timing {
+    pub ping_ms: i64,
+    pub pingreq_ms: i64,
+    pub suspicion_timeout_protocol_periods: i64,
+}
+
+impl Default for Timing {
+    fn default() -> Timing {
+        Timing {
+            ping_ms: PING_TIMING_DEFAULT_MS,
+            pingreq_ms: PINGREQ_TIMING_DEFAULT_MS,
+            suspicion_timeout_protocol_periods: SUSPICION_TIMEOUT_DEFAULT_PROTOCOL_PERIODS,
+        }
+    }
+}
+
+impl Timing {
+    /// Set up a new Timing
+    pub fn new(ping_ms: i64, pingreq_ms: i64, suspicion_timeout_protocol_periods: i64) -> Timing {
+        Timing {
+            ping_ms: ping_ms,
+            pingreq_ms: pingreq_ms,
+            suspicion_timeout_protocol_periods: suspicion_timeout_protocol_periods,
+        }
+    }
+
+    /// How long is a protocl period, in millis.
+    pub fn protocol_period_ms(&self) -> i64 {
+        self.ping_ms + self.pingreq_ms
+    }
+
+    /// When should this ping record time out?
+    pub fn ping_timeout(&self) -> SteadyTime {
+        SteadyTime::now() + TimeDuration::milliseconds(self.ping_ms)
+    }
+
+    /// When should this pingreq timeout?
+    pub fn pingreq_timeout(&self) -> SteadyTime {
+        SteadyTime::now() + TimeDuration::milliseconds(self.pingreq_ms)
+    }
+
+    /// How long before the next scheduled protocol period
+    pub fn next_protocol_period(&self) -> SteadyTime {
+        SteadyTime::now() + TimeDuration::milliseconds(self.ping_ms + self.pingreq_ms)
+    }
+
+    /// How long before this suspect entry times out
+    pub fn suspicion_timeout_duration(&self) -> TimeDuration {
+        TimeDuration::milliseconds(self.protocol_period_ms() *
+                                   self.suspicion_timeout_protocol_periods)
+    }
+}
+
+/// The outbound thread
+pub struct Outbound<'a> {
+    pub server: &'a Server,
+    pub rx_inbound: mpsc::Receiver<(SocketAddr, Swim)>,
+    pub timing: Timing,
+}
+
+impl<'a> Outbound<'a> {
+    /// Creates a new Outbound struct.
+    pub fn new(server: &'a Server,
+               rx_inbound: mpsc::Receiver<(SocketAddr, Swim)>,
+               timing: Timing)
+               -> Outbound {
+        Outbound {
+            server: server,
+            rx_inbound: rx_inbound,
+            timing: timing,
+        }
+    }
+
+    /// Run the outbound thread. Gets a list of members to pinmg, then walks the list, probing each
+    /// member.
+    ///
+    /// If the probe completes before the next protocol period is scheduled, waits for the protocol
+    /// period to finish before starting the next probe.
+    pub fn run(&mut self) {
+        loop {
+            if self.server.pause.load(Ordering::Relaxed) {
```

Comment:
Me too. It really helped my brain to think about it as integration testing rather than relying purely on functional tests.